added human mortality database service

dragazo opened the pull request:
Added a service to access the [Human Mortality Database](https://www.mortality.org/). Their data is arranged in a CSV file available for direct download from their website. I've written this service to use a meyers singleton pattern so that we automatically download and parse the file once per server startup, *but only upon request (use of the service)*. Thus, we have automatically updating data with no overhead if not being used (e.g. if we have to do a lot of server restarts to fix something unrelated).

brollb commented on src/server/services/procedures/mortality/mortality.js:
```diff
@@ -0,0 +1,228 @@
+ */
+'use strict';
+
+// source: https://www.mortality.org/
```
It would be good to put this url in the service's help message. I usually just add a line by itself and say something like "For more information, check out <url>"

brollb commented on src/server/services/procedures/mortality/mortality.js:
```diff
@@ -0,0 +1,228 @@
+ * @returns {Array}
+ */
+mortality.getDataForCountry = async function(country) {
+    return (await getData())[country] || {};
```
It might be nice to throw an error if the country is not found.

dragazo replied:
Ah, ok. In the new commit I've changed these so that they give back an error message if country does not exist, but still return empty if year and/or week don't exist.

brollb commented on src/server/services/procedures/mortality/mortality.js:
```diff
@@ -0,0 +1,228 @@
+ * If the specified country and year are not in the data, returns an empty object (no rows).
+ *
+ * @param {String} country Country to look up
+ * @param {Number} year Year to look up
```
Shouldn't year be bounded, too?

dragazo replied:
Yeah, I considered it possible they'd add data retroactively, but that's probably not very likely. In the new commit I've restricted it to 1990+ since that's as far back as it goes. I'm just hesitant to put an upper bound.

brollb commented on src/server/services/procedures/mortality/mortality.js:
```diff
@@ -0,0 +1,228 @@
+ * If the specified country, year, and week are not in the data, returns an empty object (no rows).
+ *
+ * @param {String} country Country to look up
+ * @param {Number} year Year to look up
```
Same comment as above.

brollb commented on src/server/services/procedures/mortality/mortality.js:
```diff
@@ -0,0 +1,228 @@
+ * This is an object organized by country, then by week, then broken down by gender.
+ * If the specified year is not in the data, returns an empty object (no rows).
+ *
+ * @param {Number} year Year to look up
```
Same as above. Probably should be a bounded number.

brollb commented on src/server/services/procedures/mortality/mortality.js:
```diff
@@ -0,0 +1,228 @@
+ * 
+ * @alpha
+ * @service
+ * @category Health
```
Maybe Science instead? It could be placed under "Demographic Studies" (probably nested under science) but as it would be the only service in that category, it probably should wait :)

dragazo replied:
Yeah, Science is probably the closest fit considering the other services in that category. Akos just mentioned wanting at some point to have more health-oriented data sources, but they'd still technically be Science as well. In the new commit I've changed it to Science.

brollb replied:
Gotcha. Makes sense. When I first read "Health", I was imagining connecting to devices like a fitbit or something :)

brollb reviewed (changes requested):
Nice work. I added some comments below. It also might be good to rename it to HumanMortalityDatabase to match the original data source. There are a couple services that don't follow this convention but I generally prefer the additional transparency that it provides.

Also, it looks like this is actually the "Short-term Mortality Fluctuations (STMF) data series" from the website. Does the Human Mortality Database provide any other datasets of interest?

brollb commented:
Oh, it also might be nice to rename `getEverything` to something like `getAllData`.

brollb commented:
Also, do you know how often the dataset is updated? It also might be something where we will want to update the dataset periodically assuming the server has been up for awhile.

dragazo commented:
Ok, I think I addressed everything in the new commit.

The dataset (STMF) I believe they said is updated weekly. I thought server restarts were frequent enough for that to work, but it sounds like that might not be the case. In the new commit I've changed it so that the cache is invalidated after 1 day. We could make the period longer, but the download itself is relatively lightweight, and would still only be triggered on use.

They do have other datasets, but as far as I've seen they're just more of the same. The others appear to be yearly evolutions rather than weekly, and lack some of the statistics (they have count but not rate, which appears to not just be a linear mapping). Also the others require a login to access, so we might have to deal with the usage limitations (for instance, they'd rather us link users to the data directly rather than act as an intermediary).

brollb commented on src/server/services/procedures/human-mortality-database/human-mortality-database.js:
```diff
@@ -0,0 +1,236 @@
+ * @returns {Array}
+ */
+mortality.getDataForCountry = async function(country) {
+    return (await getData())[country] || `country '${country}' is not in the database`;
```
Throw an error with the error message rather than returning the string. The error will result in the "error" block in netsblox being set appropriately.

brollb commented on src/server/services/procedures/human-mortality-database/human-mortality-database.js:
```diff
@@ -0,0 +1,236 @@
+ * @returns {Array}
+ */
+mortality.getDataForCountryAndYear = async function(country, year) {
+    const data = (await getData())[country];
```
Why not just `const data = await this.getDataForCountry(country)` here? This should save some code duplication.

dragazo replied:
Since `getDataForCountry` was returning a string on failure that was to avoid indexing the string (on failure) instead of the map (on success).

brollb commented on src/server/services/procedures/human-mortality-database/human-mortality-database.js:
```diff
@@ -0,0 +1,236 @@
+ */
+mortality.getDataForCountryYearAndWeek = async function(country, year, week) {
+    const data = (await getData())[country];
+    if (data === undefined) return `country '${country}' is not in the database`;
```
Same as above (though you could probably use `getDataForCountryAndYear`)

brollb reviewed (changes requested):
Nice. This is looking good and it should be almost ready for alpha release. There are a couple comments below about error handling.

I also noticed while playing with it a little more that each data point contains death and death rate info for a given week which I am wary could be a little unwieldy:
![DeepinScreenshot_select-area_20201118141151](https://user-images.githubusercontent.com/4982789/99582874-4f808780-29a8-11eb-9bc3-808caed0cd23.png)

Do you think it would be better to split the RPCs into `getDeathRateByCountry` and `getDeathsByCountry`, etc?

brollb commented:
Oh, I almost forgot - there should also be some interface tests at the very least.

dragazo commented:
Ok, in the latest commit I've changed a lot of the accessors and the data structure. Now, instead of separate year and week we combine them into an actual date, then format it as `mm/dd/yyyy`.  I've left the raw data accessors (but renamed `getDataForCountry` to `getAllDataForCountry`), but the most useful from a user perspective is probably the new `getTimeSeries` accessor.
It takes filter parameters and just returns a list of date/value pairs that can be used with the chart drawing service.
Country is required and has no default, but the others default to 'both' (gender) and 'deaths total' (category).
I also changed the error handling; it now correctly sets the error block in netsblox, but it uses the response status thing (if there's a better way, let me know).
Oh, and I'll probably be adding the interface tests tomorrow (seeing if there's anything else that'd be nice to add).

brollb commented:
`getTimeSeries` looks like a nice addition. I am a little concerned about the discoverability of the categories; it might be nice to add something like `getCategories` to help with that. Also, since NetsBlox is used outside of the US as well, it might be good to specify that the returned values are in "mm/dd/yyyy" format (not "dd/mm/yyyy" which is more broadly used).

As far as errors go, you should be able to just throw the error and NetsBlox should take care of the rest (so you shouldn't need to set the status code directly).

dragazo commented:
Added RPCs to get list of genders and categories. Also added the interface tests.

When I tried changing the errors to `throw 'message'` it just returns empty string to netsblox, `error` block shows empty as well, and I just get an unhandled exception on the server console (so I've left those as response status for now). Is it not supposed to work like that?

brollb commented:
Have you tried `throw new Error('message')`?

dragazo commented:
Ah, yeah throwing `Error` does work. I was just assuming it would catch anything and just stringify it.

brollb commented:
Nice work, @dragazo!
